### PR TITLE
register package in ament index

### PR DESCRIPTION
**Pull request by mikaelarguedas**
Signed-off-by: Mikael Arguedas <mikael.arguedas@gmail.com>

Installing a `package.xml` installed is not enough for rosdep to detect the package.
Making this an `ament_package` allows it to be registered in the index and recognized by rosdep. This permits to remove hacks like setting the `ROS_PACKAGE_PATH` manually for the packages to be detected.

Relates to https://github.com/ros-infrastructure/rosdep/issues/724 and https://github.com/osrf/docker_images/pull/328

### CODE REVIEW

**Comment by dirk-thomas** [comment]
> Installing a `package.xml` installed is not enough for rosdep to detect the package.

Correct, the install prefix must also be listed in the environment variable `AMENT_PREFIX_PATH` (similar to the `ROS_PACKAGE_PATH` in ROS 1).

Automatically extending environment variables from a non-ament package is certainly a problem. That is one major reason why ament exists in the first place.

Since ros-infrastructure/rosdep#699 `rosdep` also supports the ament index for faster package discovery. That might be the easier way to get this working (see next paragraph).

> Making this an `ament_package` allows it to be registered in the index and recognized by rosdep.

That change seems to be overkill just to register the package in the ament index. That registration is nothing else than installing an empty file into a specific location in the file system. The CMake logic of this package doesn't need to be fundamentally changed but only something like this be added:

```
# create an empty file with the basename ${PROJECT_NAME}, e.g. with file(WRITE ...)
install(
  FILES somewhere/${PROJECT_NAME}
  DESTINATION share/ament_index/resource_index/packages
)
```

See the logic in `ament_cmake` for reference: https://github.com/ament/ament_cmake/blob/83e1b8147e2eda7d805e45c163aea60230a4d160/ament_cmake_core/cmake/index/ament_index_register_package.cmake#L29 and https://github.com/ament/ament_cmake/blob/83e1b8147e2eda7d805e45c163aea60230a4d160/ament_cmake_core/cmake/index/ament_index_register_resource.cmake#L88

**Comment by mikaelarguedas** [comment]
> That change seems to be overkill just to register the package in the ament index. That registration is nothing else than installing an empty file into a specific location in the file system. The CMake logic of this package doesn't need to be fundamentally changed but only something like this be added:
> Since ros-infrastructure/rosdep#699 rosdep also supports the ament index for faster package discovery. That might be the easier way to get this working (see next paragraph).

That was my initial guess as well. But installing a marker file is not sufficient as the only folders considered for ament index resources are the ones listed on the AMENT_PREFIX_PATH (in [get_search_paths function](https://github.com/ros-infrastructure/rosdep/blob/fbeb43a4f76f60c16aff0545ee2caeb14bcecdd5/src/rosdep2/ament_packages/search_paths.py#L20)).
Which led me to the same statement as you: 
> Automatically extending environment variables from a non-ament package is certainly a problem. That is one major reason why ament exists in the first place.

So I ended up converting the package to be a real ament package rather than a non ament package that pretends to be one (installing package.xml, marker file in the index and modifying AMENT_PREFIX_PATH).

File: CMakeLists.txt [hunk 1]
```diff
@@ -37,16 +39,6 @@ else()
   message(STATUS "Found TinyXML. Using TinyXML from system.")
 endif()
 
-configure_file(tinyxml_vendorConfig.cmake.in "${PROJECT_BINARY_DIR}/tinyxml_vendorConfig.cmake" @ONLY)
-configure_file(tinyxml_vendorConfig-version.cmake.in "${PROJECT_BINARY_DIR}/tinyxml_vendorConfig-version.cmake" @ONLY)
```
Comment (dirk-thomas) [inline CMakeLists.txt:41]: If these templates are now obsolete please delete them as part of the PR.

**Comment by dirk-thomas** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8906/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4689/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7289/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8824/)

**Comment by dirk-thomas** [comment]
The change breaks the naming of the CMake config file which results in downstream packages being unable to find this package anymore.

**Comment by mikaelarguedas** [comment]
> The change breaks the naming of the CMake config file which results in downstream packages being unable to find this package anymore.

Sorry about that, https://github.com/ros2/tinyxml_vendor/pull/14/commits/40074721deb524aefab04e0b7412bacf12dfe795 should address this. The issue seems to be that the cmake Module provided by this package was not added to the CMake module path anymore.
I didnt notice a config file naming inconsistency, @dirk-thomas can you clarify which files you are referring to ?

---

In general, the problem I am trying to solve is for consumers of the nightly archive (and consumers of source builds in general) to be able to build overlay and use rosdep to install the dependencies of the archive and of their overlays.
It looks like the current situation is:
- ament packages can be found by rosdep
- cmake packages (mostly third-party or vendor package) need both a package.xml, be registered in the ament index and be on the AMENT_PREFIX_PATH to be found
  - no cmake package seem to modify the AMENT_PREFIX_PATH
  - most of them dont register themselves in the ament resource index

Goal: all packages are identified by rosdep with minimal changes to the third-party packages

Potential solutions:
- The current workaround is to: provide [bogus rosdep rules](https://github.com/osrf/docker_images/blob/b47cb1fdd47dbe40327a055f41dbbaa1f9b261e6/ros2/nightly/nightly/prereqs.yaml) for third-party packages  + modify the ROS_PACKAGE_PATH for non-ament packages to be detected as "catkin" packages
- The approach taken in this PR: change vendor packages to be ament packages removing the need for a ROS_PACKAGE_PATH hack. Such logic can be extended to the other packages
  - con: harder to integrate third party packages as the change is not just installing a package.xml but changing the "package structure" by making it an ament package
- Alternatives:
  - just register packages in the ament index without modifying the AMENT_PREFIX_PATH
    - package not identified by rosdep in isolated build cases
  - provide more bogus rosdep rules in the docker image
    - does not work for people using the nightly archives directly or working with source builds
  - ?

**Comment by dirk-thomas** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8952/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4720/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7322/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8914/)

**Comment by dirk-thomas** [comment]
Thanks for the patch. CI finally passed. Merging...

**Pull request merged**